psci/Main.hs

```diff
@@ -0,0 +1,349 @@
+        Just (ExitFailure _, _,   err) -> PSCI $ outputStrLn err
+        Nothing                        -> PSCI $ outputStrLn "Couldn't find node.js"
+
+-- |
+-- Takes a list of declarations and updates the environment, then run a make. If the declaration fails,
+-- restore the original environment.
+--
+handleDecls :: [P.Declaration] -> PSCI ()
+handleDecls ds = do
+  st <- get
+  let st' = updateLets ds st
+  let m = createTemporaryModule False st' (P.Literal (P.ObjectLiteral []))
+  e <- liftIO . runMake $ rebuild st' m
+  case e of
+    Left err -> PSCI $ printErrors err
+    Right _ -> put st'
+
+-- |
+-- Show actual loaded modules in psci.
+--
+handleShowLoadedModules :: PSCI ()
+handleShowLoadedModules = do
+    loadedModules <- gets psciLoadedExterns
+    PSCI $ outputStrLn (readModules loadedModules)
+  where
+    readModules = unlines . sort . nub . map (P.runModuleName . P.getModuleName . fst)
+
+-- |
+-- Show the imported modules in psci.
+--
+handleShowImportedModules :: PSCI ()
+handleShowImportedModules = do
+  PSCiState { psciImportedModules = importedModules } <- get
+  liftIO $ showModules importedModules >>= putStrLn
+  return ()
+  where
+  showModules = return . unlines . sort . map showModule
+  showModule (mn, declType, asQ) =
+    "import " ++ N.runModuleName mn ++ showDeclType declType ++
+    foldMap (\mn' -> " as " ++ N.runModuleName mn') asQ
+
+  showDeclType P.Implicit = ""
+  showDeclType (P.Explicit refs) = refsList refs
+  showDeclType (P.Hiding refs) = " hiding " ++ refsList refs
+  refsList refs = " (" ++ commaList (map showRef refs) ++ ")"
+
+  showRef :: P.DeclarationRef -> String
+  showRef (P.TypeRef pn dctors) = N.runProperName pn ++ "(" ++ maybe ".." (commaList . map N.runProperName) dctors ++ ")"
+  showRef (P.TypeOpRef ident) = "type (" ++ N.runIdent ident ++ ")"
+  showRef (P.ValueRef ident) = N.runIdent ident
+  showRef (P.TypeClassRef pn) = "class " ++ N.runProperName pn
+  showRef (P.TypeInstanceRef ident) = N.runIdent ident
+  showRef (P.ModuleRef name) = "module " ++ N.runModuleName name
+  showRef (P.PositionedDeclarationRef _ _ ref) = showRef ref
+
+  commaList :: [String] -> String
+  commaList = intercalate ", "
+
+-- |
+-- Imports a module, preserving the initial state on failure.
+--
+handleImport :: ImportedModule -> PSCI ()
+handleImport im = do
+   st <- updateImportedModules im <$> PSCI (lift get)
+   let m = createTemporaryModuleForImports st
+   e <- liftIO . runMake $ rebuild st m
+   case e of
+     Left errs -> PSCI $ printErrors errs
+     Right _  -> put st
+
+-- |
+-- Takes a value and prints its type
+--
+handleTypeOf :: P.Expr -> PSCI ()
+handleTypeOf val = do
+  st <- get
+  let m = createTemporaryModule False st val
+  e <- liftIO . runMake $ rebuild st m
+  case e of
+    Left errs -> PSCI $ printErrors errs
+    Right (_, env') ->
+      case M.lookup (P.ModuleName [P.ProperName "$PSCI"], P.Ident "it") (P.names env') of
```

Comment:
Not that it makes to big of a difference, but we should also be able to look up the value and it's type by looking at the produced `ExternsFile`. If this would also work for `handleKindOf`, `rebuild` wouldn't need to return an `Environment`